### DISCUSSION TITLE
How to modify data object set behavior (aka where are Data Object functions located?)

### DISCUSSION THREAD

**Discussion by brbintenyears** [pimcore/pimcore#10805] [Q&A / Support]
For each created data object, there are getters and setters made. Let's say, getImage() and setImage().

I have a data object that contains an Image type. On upload, I would like to add an additional step to it (convert the image to webp format then store a path to that as well in the data object's entry in DB). I already have that function ready, I'm just not sure where to place it such that it is executed on saving a modification to an instance of the data object.

**Comment by NiklasBr** [comment]
Check out the [Asset Events](https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/Event_API_and_Event_Manager.html), especially the `POST_ADD` or `POST_UPDATE`. In your event listener you should be able to get your asset:

```php
    public static function getSubscribedEvents(): array
    {
        return [
            AssetEvents::POST_ADD => 'doStuffWithAsset',
        ];
    }

    public function doStuffWithAsset(Asset $asset) {}
```

If you want to get it on events on the dataobject rather than asset, then [DataObjectEvents](https://github.com/pimcore/pimcore/blob/10.x/lib/Event/DataObjectEvents.php) are your friend.

**Comment by dvesh3** [comment]
You can override the generated class and extend the functionality like explained here https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/Overriding_Models.html#page_Overriding-Models-Entities-in-Pimcore
